# longreadASE-workflow
Workflow for allele-specific analysis in long-read data from PacBio

## ISO-seq
### Using SMRT Link
Read about SMRT Link [here](https://www.pacb.com/wp-content/uploads/SMRT_Link_User_Guide_v12.0.pdf).
The necessary files will be generated by the software.

### Using the terminal
Raw reads have to be demultiplexed and then filtered to remove primer and concatemers. To do so, the following commands have to be used:
`lima raw.reads` 
For further details refer to [pbbioconda](https://github.com/PacificBiosciences/pbbioconda) and [isoseq](https://isoseq.how/) documentations.
## Alignment
Using the flnc.bam file 

## Allele-specific Analysis

`pbmm2 align reference.fa reads.bam aligned.bam --sort -j 4 -J 2`

`whatshap haplotag -o haplotagged.bam --reference reference.fa snp.vcf aligned.bam --ignore-read-groups --ignore-linked-read --skip-missing-contigs --output-haplotag-list haplotypes.tsv`

`whatshap split --output-h1 h1.bam --output-h2 h2.bam aligned.bam haplotypes.tsv`
